File: app/src/main/java/com/termux/app/utils/NetworkErrorHandler.java
package com.termux.app.utils;

import android.content.Context;
import androidx.appcompat.app.AlertDialog;
import com.termux.shared.logger.Logger;

import java.net.ConnectException;
import java.net.NoRouteToHostException;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;
import java.util.concurrent.TimeoutException;

/**
 * 网络错误处理工具类
 * 提供统一的网络错误分类和用户友好的错误提示
 */
public class NetworkErrorHandler {
    
    private static final String LOG_TAG = "NetworkErrorHandler";
    
    /**
     * 错误类型枚举
     */
    public enum ErrorType {
        NETWORK_UNAVAILABLE,
        CONNECTION_TIMEOUT,
        HOST_UNREACHABLE,
        AUTHENTICATION_FAILED,
        PERMISSION_DENIED,
        FILE_NOT_FOUND,
        CONNECTION_LOST,
        SERVER_ERROR,
        UNKNOWN_ERROR
    }
    
    /**
     * 错误信息类
     */
    public static class ErrorInfo {
        private ErrorType type;
        private String title;
        private String message;
        private String technicalDetails;
        private boolean isRetryable;
        
        public ErrorInfo(ErrorType type, String title, String message, String technicalDetails, boolean isRetryable) {
            this.type = type;
            this.title = title;
            this.message = message;
            this.technicalDetails = technicalDetails;
            this.isRetryable = isRetryable;
        }
        
        // Getters
        public ErrorType getType() { return type; }
        public String getTitle() { return title; }
        public String getMessage() { return message; }
        public String getTechnicalDetails() { return technicalDetails; }
        public boolean isRetryable() { return isRetryable; }
    }
    
    /**
     * 分析异常并返回错误信息
     */
    public static ErrorInfo analyzeError(Throwable throwable) {
        Logger.logError(LOG_TAG, "Analyzing error: " + throwable.getClass().getSimpleName() + " - " + throwable.getMessage());
        
        String message = throwable.getMessage() != null ? throwable.getMessage() : "";
        
        // 网络连接错误
        if (throwable instanceof UnknownHostException) {
            return new ErrorInfo(
                ErrorType.HOST_UNREACHABLE,
                "无法连接到服务器",
                "请检查服务器地址是否正确，或者网络连接是否正常。",
                "DNS解析失败: " + message,
                true
            );
        }
        
        // 连接超时
        if (throwable instanceof SocketTimeoutException || throwable instanceof TimeoutException) {
            return new ErrorInfo(
                ErrorType.CONNECTION_TIMEOUT,
                "连接超时",
                "服务器响应超时，请检查网络连接或稍后重试。",
                "连接超时: " + message,
                true
            );
        }
        
        // 连接拒绝
        if (throwable instanceof ConnectException) {
            if (message.contains("Connection refused")) {
                return new ErrorInfo(
                    ErrorType.HOST_UNREACHABLE,
                    "连接被拒绝",
                    "无法连接到服务器，请检查服务器是否运行或端口是否正确。",
                    "连接被拒绝: " + message,
                    true
                );
            }
        }
        
        // 网络不可达
        if (throwable instanceof NoRouteToHostException) {
            return new ErrorInfo(
                ErrorType.NETWORK_UNAVAILABLE,
                "网络不可达",
                "无法到达目标服务器，请检查网络连接。",
                "网络路由错误: " + message,
                true
            );
        }
        
        // SSH认证失败
        if (message.toLowerCase().contains("auth") || message.toLowerCase().contains("authentication")) {
            return new ErrorInfo(
                ErrorType.AUTHENTICATION_FAILED,
                "身份验证失败",
                "用户名或密码错误，请检查登录凭据。",
                "认证失败: " + message,
                true
            );
        }
        
        // 权限拒绝
        if (message.toLowerCase().contains("permission denied") || message.toLowerCase().contains("access denied")) {
            return new ErrorInfo(
                ErrorType.PERMISSION_DENIED,
                "权限不足",
                "没有足够的权限执行此操作，请检查用户权限。",
                "权限拒绝: " + message,
                false
            );
        }
        
        // 文件不存在
        if (message.toLowerCase().contains("no such file") || message.toLowerCase().contains("file not found")) {
            return new ErrorInfo(
                ErrorType.FILE_NOT_FOUND,
                "文件不存在",
                "指定的文件或目录不存在。",
                "文件未找到: " + message,
                false
            );
        }
        
        // 连接丢失
        if (message.toLowerCase().contains("connection lost") || message.toLowerCase().contains("broken pipe")) {
            return new ErrorInfo(
                ErrorType.CONNECTION_LOST,
                "连接中断",
                "与服务器的连接已中断，请重新连接。",
                "连接丢失: " + message,
                true
            );
        }
        
        // 服务器错误
        if (message.toLowerCase().contains("server error") || message.toLowerCase().contains("internal error")) {
            return new ErrorInfo(
                ErrorType.SERVER_ERROR,
                "服务器错误",
                "服务器内部发生错误，请联系管理员或稍后重试。",
                "服务器错误: " + message,
                true
            );
        }
        
        // 未知错误
        return new ErrorInfo(
            ErrorType.UNKNOWN_ERROR,
            "未知错误",
            "发生了未知错误，请查看详细信息或联系技术支持。",
            throwable.getClass().getSimpleName() + ": " + message,
            true
        );
    }
    
    /**
     * 显示错误对话框
     */
    public static void showErrorDialog(Context context, Throwable throwable, Runnable retryAction) {
        ErrorInfo errorInfo = analyzeError(throwable);
        showErrorDialog(context, errorInfo, retryAction);
    }
    
    /**
     * 显示错误对话框
     */
    public static void showErrorDialog(Context context, ErrorInfo errorInfo, Runnable retryAction) {
        AlertDialog.Builder builder = new AlertDialog.Builder(context)
            .setTitle(errorInfo.getTitle())
            .setMessage(errorInfo.getMessage())
            .setPositiveButton("确定", null);
        
        // 如果错误可重试，添加重试按钮
        if (errorInfo.isRetryable() && retryAction != null) {
            builder.setNegativeButton("重试", (dialog, which) -> {
                Logger.logInfo(LOG_TAG, "User requested retry for error: " + errorInfo.getType());
                retryAction.run();
            });
        }
        
        // 添加查看详情按钮
        builder.setNeutralButton("详情", (dialog, which) -> {
            showTechnicalDetailsDialog(context, errorInfo);
        });
        
        builder.show();
    }
    
    /**
     * 显示技术详情对话框
     */
    private static void showTechnicalDetailsDialog(Context context, ErrorInfo errorInfo) {
        new AlertDialog.Builder(context)
            .setTitle("错误详情")
            .setMessage("错误类型: " + errorInfo.getType() + "\n\n" +
                       "技术详情: " + errorInfo.getTechnicalDetails())
            .setPositiveButton("确定", null)
            .show();
    }
    
    /**
     * 检查是否为网络相关错误
     */
    public static boolean isNetworkError(Throwable throwable) {
        return throwable instanceof UnknownHostException ||
               throwable instanceof SocketTimeoutException ||
               throwable instanceof ConnectException ||
               throwable instanceof NoRouteToHostException ||
               throwable instanceof TimeoutException;
    }
    
    /**
     * 检查是否为可重试错误
     */
    public static boolean isRetryableError(Throwable throwable) {
        ErrorInfo errorInfo = analyzeError(throwable);
        return errorInfo.isRetryable();
    }
    
    /**
     * 获取简化的错误消息
     */
    public static String getSimpleErrorMessage(Throwable throwable) {
        ErrorInfo errorInfo = analyzeError(throwable);
        return errorInfo.getMessage();
    }
}

File: app/src/main/java/com/termux/app/bridge/utils/NetworkErrorHandler.kt
package com.termux.app.bridge.utils

import java.net.ConnectException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import kotlin.math.min
import kotlin.math.pow

/**
 * 网络错误处理器
 * 专门处理网络相关的错误，提供用户友好的错误信息和重试策略
 */
object NetworkErrorHandler {
    
    /**
     * 获取用户友好的错误信息
     */
    @JvmStatic
    fun getReadableErrorMessage(error: Throwable): String {
        return when {
            error is ConnectException -> {
                "无法连接到服务器，请检查网络连接和服务器地址"
            }
            error is SocketTimeoutException -> {
                "连接超时，请检查网络状况或稍后重试"
            }
            error is UnknownHostException -> {
                "无法解析服务器地址，请检查主机名是否正确"
            }
            error.message?.contains("SSH") == true -> {
                "SSH连接失败: ${error.message}"
            }
            error.message?.contains("Authentication", ignoreCase = true) == true -> {
                "身份验证失败，请检查用户名和密码"
            }
            error.message?.contains("Permission denied", ignoreCase = true) == true -> {
                "权限被拒绝，请检查用户权限"
            }
            else -> {
                "网络错误: ${error.message ?: "未知错误"}"
            }
        }
    }
    
    /**
     * 判断是否为可重试的错误
     */
    @JvmStatic
    fun isRetryableError(error: Throwable): Boolean {
        return error is SocketTimeoutException ||
               error is ConnectException ||
               (error.message?.contains("timeout", ignoreCase = true) == true)
    }
    
    /**
     * 获取建议的重试延迟时间（毫秒）
     * 使用指数退避策略
     */
    @JvmStatic
    fun getSuggestedRetryDelay(retryCount: Int): Int {
        // 指数退避策略: 1秒, 2秒, 4秒, 8秒, 16秒, 30秒(最大)
        return min(1000 * 2.0.pow(retryCount).toInt(), 30000)
    }
    
    /**
     * 错误类型枚举
     */
    enum class ErrorType {
        NETWORK_UNAVAILABLE,
        CONNECTION_REFUSED,
        TIMEOUT,
        DNS_RESOLUTION_FAILED,
        AUTHENTICATION_FAILED,
        PERMISSION_DENIED,
        SSH_ERROR,
        UNKNOWN_ERROR
    }
    
    /**
     * 分析错误类型
     */
    @JvmStatic
    fun analyzeErrorType(error: Throwable): ErrorType {
        return when {
            error is UnknownHostException -> ErrorType.DNS_RESOLUTION_FAILED
            error is ConnectException -> ErrorType.CONNECTION_REFUSED
            error is SocketTimeoutException -> ErrorType.TIMEOUT
            error.message?.contains("Authentication", ignoreCase = true) == true -> ErrorType.AUTHENTICATION_FAILED
            error.message?.contains("Permission denied", ignoreCase = true) == true -> ErrorType.PERMISSION_DENIED
            error.message?.contains("SSH", ignoreCase = true) == true -> ErrorType.SSH_ERROR
            else -> ErrorType.UNKNOWN_ERROR
        }
    }
    
    /**
     * 获取错误修复建议
     */
    @JvmStatic
    fun getErrorSuggestion(error: Throwable): String {
        return when (analyzeErrorType(error)) {
            ErrorType.DNS_RESOLUTION_FAILED -> "请检查主机名拼写是否正确，或尝试使用IP地址连接"
            ErrorType.CONNECTION_REFUSED -> "目标服务器拒绝连接，请检查服务是否正在运行或防火墙设置"
            ErrorType.TIMEOUT -> "连接超时，请检查网络连接或尝试增加超时时间"
            ErrorType.AUTHENTICATION_FAILED -> "身份验证失败，请检查用户名、密码或SSH密钥"
            ErrorType.PERMISSION_DENIED -> "权限不足，请检查用户是否有相应的操作权限"
            ErrorType.SSH_ERROR -> "SSH连接出现问题，请检查SSH服务器配置和密钥设置"
            ErrorType.NETWORK_UNAVAILABLE -> "网络不可用，请检查网络连接状态"
            ErrorType.UNKNOWN_ERROR -> "发生未知错误，请查看详细日志信息"
        }
    }
}